### DISCUSSION TITLE
Closed symmetrical shape using splines

### DISCUSSION THREAD

**Discussion by udif** [zalo/CascadeStudio#65] [Q&A] [answered]
I'm trying to create a closed shape using BSpline(), later to be turned into a solid using Loft().

I'm having problems with the spline edges.
To demonstrate the problem I've repeated the circular spline twice so that you can see what I want vs. what I get.
I've added tiny spheres for debug so that  I can see where the split is passing relative to the points I give it.

Any idea how to achieve what I want without the split end issues, or how do I cut unwanted segments out of a split?

```
function p2xyz(r, d, z = 0) {
    let th = d / 180 * Math.PI
    let v = ([r * Math.cos(th), r * Math.sin (th), z])
    Translate(v, Rotate([0, 0, 1], d, Sphere(0.1)))
    return (v)
}

let l = []
let N=6
let r = 10
let dinc = 360/N
// The loop should be only until N, but I duplicate it to remove the boundary conditions on N
for (let i = 0; i < 2*N ; i++) {
    l.push (p2xyz(r * 1.1, i * dinc))
    l.push (p2xyz(r * 1.0, (i + 0.1) * dinc))
    l.push (p2xyz(r * 0.8, (i + 0.5) * dinc))
    l.push (p2xyz(r * 1.0, (i + 0.9) * dinc))
}
BSpline(l, true)
```

![spline](https://user-images.githubusercontent.com/809640/106255437-b6c97c00-6222-11eb-9c4d-fc23ddf5d97a.jpg)

**Comment by Irev-Dev** [comment]
Looks like the problem is that when specifying that the spline should be closed, the end gradient matches the initial gradient instead them both matching if that makes sense. I had a quick look at the definition of BSpline and then this https://dev.opencascade.org/doc/refman/html/class_geom_a_p_i___points_to_b_spline.html to see if there were any easy wins, but nothing struck me.

**Comment by udif** [comment, marked as answer]
I solved my problem by taking only half the spline and duplicating it:
```
function p2xyz(r, d, z = 0) {
    let th = d / 180 * Math.PI
    let v = ([r * Math.cos(th), r * Math.sin (th), z])
    //Translate(v, Rotate([0, 0, 1], d, Sphere(0.1)))
    return (v)
}

let l = []
let N=6
let r = 10
let dinc = 360/N
for (let i = 0; i < N ; i++) {
    l.push (p2xyz(r * 1.1, i * dinc))
    l.push (p2xyz(r * 1.0, (i + 0.1) * dinc))
    l.push (p2xyz(r * 0.8, (i + 0.5) * dinc))
    l.push (p2xyz(r * 1.0, (i + 0.9) * dinc))
}
base = BSpline(l, true)
top = Translate([0, 0, 100], Circle(10, true))
shape1 = Loft([base, top])
cut1 = Translate([-20, -20, 0], Box(20, 40, 100, false))
semi = Intersection([shape1, cut1])
final = Translate([100, 0, 0], Union([semi, Rotate([0, 0, 1], 180, semi)]))
![spline](https://user-images.githubusercontent.com/809640/106364146-87576400-6335-11eb-87cf-04e77ec1bb3e.jpg)
```
Screenshot is taken below the object to show the spline-based shape.

![spline](https://user-images.githubusercontent.com/809640/106364163-a7872300-6335-11eb-8c26-d59b6c246377.jpg)
